File: app/user/internal/service/user/address.go
package user

import (
	"context"

	v1 "github.com/tmjwjx/supermarket/api/user/v1"
	bizuser "github.com/tmjwjx/supermarket/app/user/internal/biz/user"
)

type AddressService struct {
	v1.UnimplementedAddressServiceServer
	uc *bizuser.AddressUsecase
}

func NewAddressService(uc *bizuser.AddressUsecase) *AddressService {
	return &AddressService{uc: uc}
}

func (s *AddressService) CreateAddress(ctx context.Context, req *v1.CreateAddressRequest) (*v1.CreateAddressResponse, error) {
	uid, err := callerID(ctx)
	if err != nil {
		return nil, err
	}
	created, err := s.uc.Create(ctx, &bizuser.Address{
		UserID: uid, Receiver: req.GetReceiver(), Phone: req.GetPhone(),
		Province: req.GetProvince(), City: req.GetCity(), District: req.GetDistrict(), Detail: req.GetDetail(),
	})
	if err != nil {
		return nil, err
	}
	return &v1.CreateAddressResponse{Address: toAddress(created)}, nil
}

func (s *AddressService) ListAddresses(ctx context.Context, _ *v1.ListAddressesRequest) (*v1.ListAddressesResponse, error) {
	uid, err := callerID(ctx)
	if err != nil {
		return nil, err
	}
	rows, err := s.uc.List(ctx, uid)
	if err != nil {
		return nil, err
	}
	out := &v1.ListAddressesResponse{}
	for _, row := range rows {
		out.Addresses = append(out.Addresses, toAddress(row))
	}
	return out, nil
}

func (s *AddressService) DeleteAddress(ctx context.Context, req *v1.DeleteAddressRequest) (*v1.DeleteAddressResponse, error) {
	uid, err := callerID(ctx)
	if err != nil {
		return nil, err
	}
	id, err := parseUserID(req.GetId())
	if err != nil {
		return nil, err
	}
	if err := s.uc.Delete(ctx, uid, id); err != nil {
		return nil, err
	}
	return &v1.DeleteAddressResponse{}, nil
}

func (s *AddressService) UpdateAddress(ctx context.Context, req *v1.UpdateAddressRequest) (*v1.UpdateAddressResponse, error) {
	uid, err := callerID(ctx)
	if err != nil {
		return nil, err
	}
	id, err := parseUserID(req.GetId())
	if err != nil {
		return nil, err
	}
	updated, err := s.uc.Update(ctx, &bizuser.Address{
		ID: id, UserID: uid, Receiver: req.GetReceiver(), Phone: req.GetPhone(),
		Province: req.GetProvince(), City: req.GetCity(), District: req.GetDistrict(), Detail: req.GetDetail(),
	})
	if err != nil {
		return nil, err
	}
	return &v1.UpdateAddressResponse{Address: toAddress(updated)}, nil
}

func (s *AddressService) SetDefaultAddress(ctx context.Context, req *v1.SetDefaultAddressRequest) (*v1.SetDefaultAddressResponse, error) {
	uid, err := callerID(ctx)
	if err != nil {
		return nil, err
	}
	id, err := parseUserID(req.GetId())
	if err != nil {
		return nil, err
	}
	updated, err := s.uc.SetDefault(ctx, uid, id)
	if err != nil {
		return nil, err
	}
	return &v1.SetDefaultAddressResponse{Address: toAddress(updated)}, nil
}

func (s *AddressService) GetAddress(ctx context.Context, req *v1.GetAddressRequest) (*v1.GetAddressResponse, error) {
	uid, err := parseUserID(req.GetUserId())
	if err != nil {
		return nil, err
	}
	id, err := parseUserID(req.GetId())
	if err != nil {
		return nil, err
	}
	row, err := s.uc.Get(ctx, uid, id)
	if err != nil {
		return nil, err
	}
	return &v1.GetAddressResponse{Address: toAddress(row)}, nil
}

func toAddress(in *bizuser.Address) *v1.Address {
	if in == nil {
		return nil
	}
	return &v1.Address{
		Id: in.ID.String(), Receiver: in.Receiver, Phone: in.Phone, Province: in.Province,
		City: in.City, District: in.District, Detail: in.Detail, IsDefault: in.IsDefault,
	}
}

File: frontend/web/src/api/address.ts
import { send } from './client.ts'
import { readBool, readList, readString } from './read.ts'

export type Address = {
  id: string
  receiver: string
  phone: string
  province: string
  city: string
  district: string
  detail: string
  isDefault: boolean
}

export function listAddresses() {
  return send('/v1/addresses', {})
}

export function createAddress(input: {
  receiver: string
  phone: string
  province: string
  city: string
  district: string
  detail: string
}) {
  return send('/v1/addresses', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
}

export function updateAddress(id: string, input: {
  receiver: string
  phone: string
  province: string
  city: string
  district: string
  detail: string
}) {
  return send(`/v1/addresses/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
}

export function deleteAddress(id: string) {
  return send(`/v1/addresses/${encodeURIComponent(id)}`, { method: 'DELETE' })
}

export function setDefaultAddress(id: string) {
  return send(`/v1/addresses/${encodeURIComponent(id)}:default`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: '{}',
  })
}

export function readAddresses(body: unknown): Address[] {
  return readList(body, 'addresses').flatMap((row) => {
    const id = readString(row, 'id')
    if (!id) return []
    return [
      {
        id,
        receiver: readString(row, 'receiver'),
        phone: readString(row, 'phone'),
        province: readString(row, 'province'),
        city: readString(row, 'city'),
        district: readString(row, 'district'),
        detail: readString(row, 'detail'),
        isDefault: readBool(row, 'is_default', 'isDefault'),
      },
    ]
  })
}
